iocore/net/quic/QUICCongestionController.cc

```diff
@@ -59,34 +60,65 @@ QUICCongestionController::on_packet_sent(size_t bytes_sent)
 }
 
 bool
-QUICCongestionController::_in_recovery(QUICPacketNumber packet_number)
+QUICCongestionController::_in_recovery(ink_hrtime sent_time)
 {
-  return packet_number <= this->_end_of_recovery;
+  return sent_time <= this->_recovery_start_time;
 }
 
 void
 QUICCongestionController::on_packet_acked(const PacketInfo &acked_packet)
 {
   // Remove from bytes_in_flight.
   SCOPED_MUTEX_LOCK(lock, this->_cc_mutex, this_ethread());
-  this->_bytes_in_flight -= acked_packet.bytes;
-  if (this->_in_recovery(acked_packet.packet_number)) {
+  this->_bytes_in_flight -= acked_packet.sent_bytes;
+  if (this->_in_recovery(acked_packet.time_sent)) {
     // Do not increase congestion window in recovery period.
     return;
   }
   if (this->_congestion_window < this->_ssthresh) {
     // Slow start.
-    this->_congestion_window += acked_packet.bytes;
+    this->_congestion_window += acked_packet.sent_bytes;
     QUICCCDebug("slow start window chaged");
   } else {
     // Congestion avoidance.
-    this->_congestion_window += this->_k_default_mss * acked_packet.bytes / this->_congestion_window;
+    this->_congestion_window += this->_k_max_datagram_size * acked_packet.sent_bytes / this->_congestion_window;
     QUICCCDebug("Congestion avoidance window changed");
   }
 }
 
 void
-QUICCongestionController::on_packets_lost(const std::map<QUICPacketNumber, PacketInfo *> &lost_packets)
+QUICCongestionController::_congestion_event(ink_hrtime sent_time, uint32_t pto_count)
+{
+  // Start a new congestion event if the sent time is larger
+  // than the start time of the previous recovery epoch.
+  if (!this->_in_recovery(sent_time)) {
+    this->_recovery_start_time = Thread::get_hrtime();
+    this->_congestion_window *= this->_k_loss_reduction_factor;
+    this->_congestion_window = std::max(this->_congestion_window, this->_k_minimum_window);
+    this->_ssthresh          = this->_congestion_window;
+    if (pto_count > this->_k_persistent_congestion_threshold) {
+      this->_congestion_window = this->_k_minimum_window;
+    }
+  }
+}
+
+void
+QUICCongestionController::process_ecn(const PacketInfo &acked_largest_packet, const QUICAckFrame::EcnSection *ecn_section,
+                                      uint32_t pto_count)
```

Comment:
```
      ProcessECN(ack):
        // If the ECN-CE counter reported by the peer has increased,
        // this could be a new congestion event.
        if (ack.ce_counter > ecn_ce_counter):
          ecn_ce_counter = ack.ce_counter
          // Start a new congestion event if the last acknowledged
          // packet was sent after the start of the previous
          // recovery epoch.
          CongestionEvent(sent_packets[ack.largest_acked].time_sent)
```

The same as comment above.

Reply:
Add ecn process

Reply:
Thanks.